provider/cluster_resource.go

```diff
@@ -39,6 +47,98 @@ type ClusterResource struct {
 	collection *cmv1.ClustersClient
 }
 
+type ClusterClient interface {
+	Get(ctx context.Context, id string) (*cmv1.Cluster, error)
+	Create(ctx context.Context, cluster *cmv1.Cluster) (*cmv1.Cluster, error)
+	PollReady(ctx context.Context, id string) error
+	PollRemoved(ctx context.Context, id string) error
+	Update(ctx context.Context, id string, cluster *cmv1.Cluster) (*cmv1.Cluster, error)
+	Delete(ctx context.Context, id string) error
+}
+
+type ClusterClientImpl struct {
+	InternalClient *cmv1.ClustersClient
+}
+
+func (clusterClient *ClusterClientImpl) Get(ctx context.Context, id string) (*cmv1.Cluster, error) {
+	response, err := clusterClient.InternalClient.Cluster(id).Get().SendContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return response.Body(), nil
+}
+
+func (clusterClient *ClusterClientImpl) Create(ctx context.Context, cluster *cmv1.Cluster) (*cmv1.Cluster, error) {
+	response, err := clusterClient.InternalClient.Add().Body(cluster).SendContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return response.Body(), nil
+}
+
+func (clusterClient *ClusterClientImpl) PollReady(ctx context.Context, id string) error {
+	pollCtx, cancel := context.WithTimeout(ctx, 1*time.Hour)
+	defer cancel()
+	_, err := clusterClient.InternalClient.Cluster(id).Poll().
+		Interval(30 * time.Second).
+		Predicate(func(get *cmv1.ClusterGetResponse) bool {
+			object := get.Body()
+			return object.State() == cmv1.ClusterStateReady
+		}).
+		StartContext(pollCtx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (clusterClient *ClusterClientImpl) PollRemoved(ctx context.Context, id string) error {
+	pollCtx, cancel := context.WithTimeout(ctx, 10*time.Minute)
+	defer cancel()
+	_, err := clusterClient.InternalClient.Cluster(id).Poll().
+		Interval(30 * time.Second).
+		Status(http.StatusNotFound).
+		StartContext(pollCtx)
+	sdkErr, ok := err.(*errors.Error)
+	if ok && sdkErr.Status() == http.StatusNotFound {
+		err = nil
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (clusterClient *ClusterClientImpl) Update(ctx context.Context, id string, patch *cmv1.Cluster) (*cmv1.Cluster, error) {
+	update, err := clusterClient.InternalClient.Cluster(id).Update().
+		Body(patch).
+		SendContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return update.Body(), nil
+}
+
+func (clusterClient *ClusterClientImpl) Delete(ctx context.Context, id string) error {
+	_, err := clusterClient.InternalClient.Cluster(id).Delete().SendContext(ctx)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type ClusterResourceUtils interface {
+	populateClusterState(object *cmv1.Cluster, state *ClusterState)
+	createClusterObject(ctx context.Context,
+		state *ClusterState, diags diag.Diagnostics) (*cmv1.Cluster, error)
+}
+
+type ClusterResourceUtilsImpl struct {
+	clusterClient ClusterClient
+}
+
```

Comment:
Also this new Interface and its implementation should be moved to different package/File